### PR TITLE
[mupdf] Add custom allocator to avoid having fuzzer kill process (#1830)

### PR DESCRIPTION
**Pull request by sebras**
This fixes oss-fuzz #5679 and oss-fuzz #7803 for the mupdf project.

I have tested this locally and works fine; detecting 
allocations that in total are above 1Gbyte.

### CODE REVIEW

File: projects/mupdf/pdf_fuzzer.cc [hunk 1]
```diff
@@ -51,5 +149,7 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
 
   fz_drop_context(ctx);
 
+  fprintf(stderr, "total allocation at end of execution: %" PRIu64 " bytes\n", total);
```
Comment (Dor1s) [inline projects/mupdf/pdf_fuzzer.cc]: It's recommended not to print anything from a fuzz target, especially inside LLVMFuzzerTestOneInput function which is executed a huge number of times.

File: projects/mupdf/pdf_fuzzer.cc [hunk 2]
```diff
@@ -17,11 +17,101 @@
...
+fz_malloc_ossfuzz(void *opaque, size_t size)
+{
+	char *ptr = NULL;
+
+	if (size == 0)
+		return NULL;
+	if (size > SIZE_MAX - ALIGNMENT)
+		return NULL;
+
+	if (total + size + ALIGNMENT > MAX_ALLOCATION)
```
Comment (oliverchang) [inline projects/mupdf/pdf_fuzzer.cc]: this addition can potentially overflow and cause the check to incorrectly fail. should this be properly handled here?

File: projects/mupdf/pdf_fuzzer.cc [hunk 3]
```diff
@@ -17,11 +17,101 @@
...
+		fz_free_ossfuzz(opaque, old);
+		return NULL;
+	}
+	if (size > SIZE_MAX - ALIGNMENT)
+		return NULL;
+
+	old = ((char *) old) - ALIGNMENT;
+	memcpy(&oldsize, old, sizeof(oldsize));
+
+	if (total - oldsize + size > MAX_ALLOCATION)
```
Comment (oliverchang) [inline projects/mupdf/pdf_fuzzer.cc]: same here

**Comment by sebras** [comment]
Good point. I have fixed those potential sources of overflow I believe. Additionally the code was being ambivalent about whether to include `ALIGNMENT` in `total` or not (it did in comparisons but not when increasing/decreasing `total`). I fixed that too. I re-tested on the reproducers from oss-fuzz #5679 and oss-fuzz #7803 locally and the allocator returns `NULL` before the fuzzer aborts the test, just like intended.

**Comment by oliverchang** [comment]
Great, thanks! Let's give this a shot.

**Pull request merged**